Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.cpp

```diff
@@ -0,0 +1,192 @@
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "CoordinatedPlatformLayerBufferVideo.h"
+
+#if USE(COORDINATED_GRAPHICS) && ENABLE(VIDEO) && USE(GSTREAMER)
+#include "CoordinatedPlatformLayerBufferExternalOES.h"
+#include "CoordinatedPlatformLayerBufferRGB.h"
+#include "CoordinatedPlatformLayerBufferYUV.h"
+#include "TextureMapper.h"
+
+// Include the <epoxy/gl.h> header before <gst/gl/gl.h>.
+#include <epoxy/gl.h>
+#include <gst/gl/gl.h>
+
+namespace WebCore {
+
+std::unique_ptr<CoordinatedPlatformLayerBufferVideo> CoordinatedPlatformLayerBufferVideo::create(GstSample* sample, std::optional<GstVideoDecoderPlatform> videoDecoderPlatform, bool gstGLEnabled, OptionSet<TextureMapperFlags> flags)
+{
+    GstVideoInfo videoInfo;
+    if (UNLIKELY(!getSampleVideoInfo(sample, videoInfo)))
+        return nullptr;
+
+    auto* buffer = gst_sample_get_buffer(sample);
+    if (UNLIKELY(!GST_IS_BUFFER(buffer)))
+        return nullptr;
+
+    return makeUnique<CoordinatedPlatformLayerBufferVideo>(buffer, &videoInfo, videoDecoderPlatform, gstGLEnabled, flags);
+}
+
+CoordinatedPlatformLayerBufferVideo::CoordinatedPlatformLayerBufferVideo(GstBuffer* buffer, GstVideoInfo* videoInfo, std::optional<GstVideoDecoderPlatform> videoDecoderPlatform, bool gstGLEnabled, OptionSet<TextureMapperFlags> flags)
+    : CoordinatedPlatformLayerBuffer(Type::Video, IntSize(GST_VIDEO_INFO_WIDTH(videoInfo), GST_VIDEO_INFO_HEIGHT(videoInfo)), flags, nullptr)
+    , m_gstBuffer(buffer)
+    , m_videoDecoderPlatform(videoDecoderPlatform)
+    , m_buffer(createBufferIfNeeded(videoInfo, gstGLEnabled))
+{
+}
+
+CoordinatedPlatformLayerBufferVideo::~CoordinatedPlatformLayerBufferVideo()
+{
+    if (m_isMapped)
+        gst_video_frame_unmap(&m_videoFrame);
+}
+
+std::unique_ptr<CoordinatedPlatformLayerBuffer> CoordinatedPlatformLayerBufferVideo::copyBuffer() const
+{
+    if (!m_buffer || !is<CoordinatedPlatformLayerBufferRGB>(*m_buffer))
+        return nullptr;
+
+    auto& buffer = downcast<CoordinatedPlatformLayerBufferRGB>(*m_buffer);
+    auto textureID = buffer.textureID();
+    if (!textureID)
+        return nullptr;
+
+    auto texture = BitmapTexture::create(buffer.size());
+    texture->copyFromExternalTexture(textureID);
+    return CoordinatedPlatformLayerBufferRGB::create(WTFMove(texture), m_flags, nullptr);
+}
+
+std::unique_ptr<CoordinatedPlatformLayerBuffer> CoordinatedPlatformLayerBufferVideo::createBufferIfNeeded(GstVideoInfo* videoInfo, bool gstGLEnabled)
+{
```

Comment:
I have a doubt here. This function is called on the constructor to initialize m_buffer. Why does it do it only for the gl case? When it's not gl, why are you waiting to paintToTextureMapper to do so? Is it with the intention of performing the gl upload during the composition?

Reply:
Yes, exactly, when we have a texture already we can just create the buffer that will be used directly in paintToTextureMapper. When we need to map the buffer and upload the pixels, we want to do the upload in the compositor thread, which also allows us to use the texture mapper pool.

Reply:
I haven't checked recently, but I think that pushTextureToCompositor is already being called in the compositor thread when using the non gstgl path (but I many be wrong and that's not the meaning of m_usingFallbackVideoSink). This would mean that you are already on the compositor thread when calling createBufferIfNeeded in the non gstgl case. Anyway, as you rearranged the code, the call to scheduleOnCompositorThread may not be needed anymore, as we wait for the actual composition to copy the buffer into a texture, unlike before.